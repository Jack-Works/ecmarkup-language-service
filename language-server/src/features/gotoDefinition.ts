import { type Connection, type TextDocuments, type ServerCapabilities, Location } from 'vscode-languageserver'
import { TextDocument } from '../lib.js'
import { getSourceFile } from '../utils/parse.js'
import { expandWord } from '../utils/text.js'
import { createRange } from '../utils/utils.js'

export function definitionProvider(
    connection: Connection,
    documents: TextDocuments<TextDocument>,
): NonNullable<ServerCapabilities['definitionProvider']> {
    connection.onDefinition(async (params, token, workDoneProgress, resultProgress) => {
        const document = documents.get(params.textDocument.uri)
        if (!document) return undefined
        const sourceFile = getSourceFile.get(document)

        const offset = document.offsetAt(params.position)
        let { word, isGrammarLeading, isVariableLeading } = expandWord(document.getText(), offset)

        const node = sourceFile.findNodeAt(offset)
        if (node.tag === 'emu-grammar' || isGrammarLeading) {
            const result = sourceFile.getGrammarDefinition(word)
            return result.map((range): Location => ({ uri: document.uri, range: range[1] }))
        } else if (node.tag === 'emu-alg') {
            if (isVariableLeading) {
                const header = sourceFile.getAlgHeader(offset)
                const headerText = sourceFile.getNodeText(header)
                const headerDefine = headerText?.indexOf(`_${word}_`)
                if (header && headerDefine && headerDefine !== -1) {
                    const headerStart = header.startTagEnd || header.start
                    return {
                        uri: document.uri,
                        range: createRange(
                            sourceFile.text.positionAt(headerStart + headerDefine),
                            sourceFile.text.positionAt(
                                (header.startTagEnd || header.start) + headerDefine + word.length + 2,
                            ),
                        ),
                    }
                }

                const nodeText = sourceFile.getNodeText(node)
                let nodeDefine = nodeText.indexOf(`Let _${word}_`)
                if (nodeDefine === -1) {
                    try {
                        const match = nodeText.match(new RegExp(`For each (\\w+) _${word}_ of `))
                        if (match) nodeDefine = match.index! + match[1]!.length + 6
                    } catch {}
                }
                if (nodeDefine === -1) return undefined
                return {
                    uri: document.uri,
                    range: createRange(
                        sourceFile.text.positionAt((node.startTagEnd || node.start) + nodeDefine + 4),
                        sourceFile.text.positionAt((node.startTagEnd || node.start) + nodeDefine + word.length + 6),
                    ),
                }
            }
        }

        return undefined
    })
    return {}
}
