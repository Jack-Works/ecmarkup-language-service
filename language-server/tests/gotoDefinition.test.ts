import { expect, it } from 'vitest'
import { GoToDefinition } from '../src/features/gotoDefinition.js'
import { File } from './File.js'

it('go to definition', async () => {
    const goto = new GoToDefinition()
    const { document, markers, textDocument, program } = File.of`
        <emu-grammar type="definition">
            MoreOneNight ::= "more" "one" "night"
        </emu-grammar>
        <p>A |MoreOneNight${File.mark('grammar')}| production.</p>
        <emu-grammar>
            Other
                MoreOneNight${File.mark('grammar in grammar')}
        </emu-grammar>

        <emu-clause id="sec-SaveTheWorld" type="abstract operation">
            <h1>
                SaveTheWorld (
                    _input_: an ECMAScript language value,
                ): either a normal completion containing an Object or a throw completion
            </h1>
            <dl class="header">
            </dl>
            <emu-alg>
                1. Let _obj_ be ! OrdinaryObjectCreate(_input_${File.mark('parameter')}).
                1. Perform ? FinishSavingTheWorld(_obj_${File.mark('local variable')}).
            </emu-alg>
        </emu-clause>
        <emu-alg>
            1. Perform ? SaveThe${File.mark('AO')}World(  ).
        </emu-alg>
    `

    for (const { position, around, desc } of markers) {
        const result = goto.onDefinition(document, program, { textDocument, position })
        expect([around, result]).toMatchSnapshot(desc)
    }
})
