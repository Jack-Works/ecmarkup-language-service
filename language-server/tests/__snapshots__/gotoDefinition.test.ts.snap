// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`go to definition > AO 1`] = `
[
  "1. Perform ? SaveTheðŸ”½World(  ).",
  [
    {
      "range": {
        "end": {
          "character": 28,
          "line": 13,
        },
        "start": {
          "character": 16,
          "line": 13,
        },
      },
      "uri": "test://1ba2f12a6737ed2ae48ad841847a364e.emu",
    },
  ],
]
`;

exports[`go to definition > emu-prodref 1`] = `
[
  "<emu-prodref name="MoreOneNightðŸ”½"></emu-prodref>",
  undefined,
]
`;

exports[`go to definition > grammar 1`] = `
[
  "<p>A |MoreOneNightðŸ”½| production.</p>",
  [
    {
      "range": {
        "end": {
          "character": 24,
          "line": 2,
        },
        "start": {
          "character": 12,
          "line": 2,
        },
      },
      "uri": "test://1ba2f12a6737ed2ae48ad841847a364e.emu",
    },
  ],
]
`;

exports[`go to definition > grammar in grammar 1`] = `
[
  "MoreOneNightðŸ”½",
  [
    {
      "range": {
        "end": {
          "character": 24,
          "line": 2,
        },
        "start": {
          "character": 12,
          "line": 2,
        },
      },
      "uri": "test://1ba2f12a6737ed2ae48ad841847a364e.emu",
    },
  ],
]
`;

exports[`go to definition > linkSupport: AO 1`] = `
[
  "1. Perform ? SaveTheðŸ”½World(  ).",
  [
    {
      "originSelectionRange": {
        "end": {
          "character": 37,
          "line": 25,
        },
        "start": {
          "character": 25,
          "line": 25,
        },
      },
      "targetRange": {
        "end": {
          "character": 88,
          "line": 15,
        },
        "start": {
          "character": 16,
          "line": 13,
        },
      },
      "targetSelectionRange": {
        "end": {
          "character": 28,
          "line": 13,
        },
        "start": {
          "character": 16,
          "line": 13,
        },
      },
      "targetUri": "test://1ba2f12a6737ed2ae48ad841847a364e.emu",
    },
  ],
]
`;

exports[`go to definition > linkSupport: emu-prodref 1`] = `
[
  "<emu-prodref name="MoreOneNightðŸ”½"></emu-prodref>",
  undefined,
]
`;

exports[`go to definition > linkSupport: grammar 1`] = `
[
  "<p>A |MoreOneNightðŸ”½| production.</p>",
  [
    {
      "originSelectionRange": {
        "end": {
          "character": 26,
          "line": 4,
        },
        "start": {
          "character": 14,
          "line": 4,
        },
      },
      "targetRange": {
        "end": {
          "character": 49,
          "line": 2,
        },
        "start": {
          "character": 12,
          "line": 2,
        },
      },
      "targetSelectionRange": {
        "end": {
          "character": 24,
          "line": 2,
        },
        "start": {
          "character": 12,
          "line": 2,
        },
      },
      "targetUri": "test://1ba2f12a6737ed2ae48ad841847a364e.emu",
    },
  ],
]
`;

exports[`go to definition > linkSupport: grammar in grammar 1`] = `
[
  "MoreOneNightðŸ”½",
  [
    {
      "originSelectionRange": {
        "end": {
          "character": 28,
          "line": 8,
        },
        "start": {
          "character": 16,
          "line": 8,
        },
      },
      "targetRange": {
        "end": {
          "character": 49,
          "line": 2,
        },
        "start": {
          "character": 12,
          "line": 2,
        },
      },
      "targetSelectionRange": {
        "end": {
          "character": 24,
          "line": 2,
        },
        "start": {
          "character": 12,
          "line": 2,
        },
      },
      "targetUri": "test://1ba2f12a6737ed2ae48ad841847a364e.emu",
    },
  ],
]
`;

exports[`go to definition > linkSupport: local variable 1`] = `
[
  "1. Perform ? FinishSavingTheWorld(_obj_ðŸ”½).",
  [
    {
      "originSelectionRange": {
        "end": {
          "character": 54,
          "line": 21,
        },
        "start": {
          "character": 51,
          "line": 21,
        },
      },
      "targetRange": {
        "end": {
          "character": 18,
          "line": 20,
        },
        "start": {
          "character": 15,
          "line": 20,
        },
      },
      "targetSelectionRange": {
        "end": {
          "character": 18,
          "line": 20,
        },
        "start": {
          "character": 15,
          "line": 20,
        },
      },
      "targetUri": "test://1ba2f12a6737ed2ae48ad841847a364e.emu",
    },
  ],
]
`;

exports[`go to definition > linkSupport: parameter 1`] = `
[
  "1. Let _obj_ be ! OrdinaryObjectCreate(_input_ðŸ”½).",
  [
    {
      "originSelectionRange": {
        "end": {
          "character": 61,
          "line": 20,
        },
        "start": {
          "character": 56,
          "line": 20,
        },
      },
      "targetRange": {
        "end": {
          "character": 22,
          "line": 14,
        },
        "start": {
          "character": 17,
          "line": 14,
        },
      },
      "targetSelectionRange": {
        "end": {
          "character": 22,
          "line": 14,
        },
        "start": {
          "character": 17,
          "line": 14,
        },
      },
      "targetUri": "test://1ba2f12a6737ed2ae48ad841847a364e.emu",
    },
  ],
]
`;

exports[`go to definition > local variable 1`] = `
[
  "1. Perform ? FinishSavingTheWorld(_obj_ðŸ”½).",
  [
    {
      "range": {
        "end": {
          "character": 18,
          "line": 20,
        },
        "start": {
          "character": 15,
          "line": 20,
        },
      },
      "uri": "test://1ba2f12a6737ed2ae48ad841847a364e.emu",
    },
  ],
]
`;

exports[`go to definition > parameter 1`] = `
[
  "1. Let _obj_ be ! OrdinaryObjectCreate(_input_ðŸ”½).",
  [
    {
      "range": {
        "end": {
          "character": 22,
          "line": 14,
        },
        "start": {
          "character": 17,
          "line": 14,
        },
      },
      "uri": "test://1ba2f12a6737ed2ae48ad841847a364e.emu",
    },
  ],
]
`;
